lib/class-apc-cache-interceptor.php

```diff
@@ -0,0 +1,494 @@
+
+			// Obtain the resolved cache key
+			$temp_flag = $flag;
+			$key       = $wp_object_cache->key( $id, $temp_flag );
+
+			// Serve from the thread cache instead of hitting APC if we've
+			// dealt with this value before during this execution
+			$thread_cache_key = sprintf( '%s:%s', $intercept['resolved']['group'], $intercept['resolved']['key'] );
+			if ( isset( $this->cache[ $thread_cache_key ] ) ) {
+				$rval = $this->maybe_clone( $this->cache[ $thread_cache_key ] );
+				$this->run_callbacks(
+					'cache_hit',
+					array(
+						'cache'     => 'thread',
+						'intercept' => $intercept,
+						'value'     => $rval,
+					)
+				);
+				return $rval;
+			}
+
+			// Attempt to fetch from acpu
+			$success = false;
+			$rval    = apcu_fetch( $key, $success );
+
+			// On success simply serve that value
+			if ( $success ) {
+				// Store in the thread cache
+				$this->cache[ $thread_cache_key ] = unserialize( $rval );
+
+				$rval = $this->maybe_clone( $this->cache[ $thread_cache_key ] );
+
+				$this->run_callbacks(
+					'cache_hit',
+					array(
+						'cache'     => 'apcu',
+						'intercept' => $intercept,
+						'value'     => $rval,
+					)
+				);
+				return $rval;
+			}
+
+			// Attempt to fetch the object from memcached
+			$rval = $wp_object_cache->get( $id, $flag, $force );
+
+			if ( false !== $rval ) {
+				// Store in the thread cache and store it in apcu
+				$this->cache[ $thread_cache_key ] = $this->maybe_clone( $rval );
+				if ( apcu_store( $key, serialize( $this->cache[ $thread_cache_key ] ), $intercept['model']['ttl'] ) ) {
+					$this->run_callbacks(
+						'cache_miss',
+						array(
+							'found'     => true,
+							'stored'    => true,
+							'intercept' => $intercept,
+							'value'     => $rval,
+						)
+					);
+				} else {
+					$this->run_callbacks(
+						'cache_miss',
+						array(
+							'found'     => true,
+							'stored'    => false,
+							'intercept' => $intercept,
+							'value'     => $rval,
+						)
+					);
+				}
+			} else {
+					$this->run_callbacks(
+						'cache_miss',
+						array(
+							'found'     => false,
+							'stored'    => false,
+							'intercept' => $intercept,
+							'value'     => $rval,
+						)
+					);
+			}
+
+			// Finally return the value we got from the object cache
+			return $rval;
+		}
+
+		private function maybe_clone( $value ) {
+			// Concept borrowed from object-cache.php itself
+			if ( is_object( $value ) ) {
+				return clone $value;
+			}
+			return $value;
+		}
+
+		public function __construct() {
+			// Allow for the possibility that this is already defined
+			if ( ! defined( 'WP_OBJ_CACHE_HOOKS' ) ) {
+				define( 'WP_OBJ_CACHE_HOOKS', true );
+			}
+			// If we should not be hooking then there's no reason to add overhead to actions and filters
+			if ( defined( 'WP_OBJ_CACHE_HOOKS' ) && WP_OBJ_CACHE_HOOKS ) {
+				do_action( 'apcu_hot_cache_init', $this );
+				add_filter( 'wp_cache_get_pre', array( $this, 'wp_cache_get' ), 1, 4 );
```

Comment:
Now that we changed the filter in https://github.com/Automattic/wp-memcached/pull/66 we should also update its usage here.
```suggestion
				add_filter( 'pre_wp_cache_get', array( $this, 'wp_cache_get' ), 1, 4 );
```